File: examples/bun/demo_streaming.ts
/**
 * mp4decrypt-api - Custom Streaming / Callback I/O Demo (Bun)
 * ============================================================
 *
 * Demonstrates Paradigm 3: Using user-defined C callbacks (wrapped in
 * TypeScript) to decrypt directly between custom stream interfaces
 * without using the filesystem or static buffers.
 *
 * Run with:
 *   bun examples/bun/demo_streaming.ts
 */

import { Mp4DecryptSession, Mp4LogLevel } from "../../sdk/bun/mp4decrypt_sdk";

async function main() {
  console.log("===================================================================");
  console.log("    MP4DECRYPT-API - CUSTOM STREAMING CALLBACK I/O DEMO (BUN)      ");
  console.log("===================================================================");

  const baseUrl = "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey";
  const kidHex = "9eb4050de44b4802932e27d75083e266";
  const keyHex = "166634c675823c235a4a9446fad52e4d";

  console.log("Fetching encrypted media chunks...");
  const [initResp, segResp] = await Promise.all([
    fetch(`${baseUrl}/15/init.mp4`),
    fetch(`${baseUrl}/15/0001.m4s`),
  ]);
  const initBytes = new Uint8Array(await initResp.arrayBuffer());
  const segBytes = new Uint8Array(await segResp.arrayBuffer());
  const inputData = new Uint8Array(initBytes.byteLength + segBytes.byteLength);
  inputData.set(initBytes, 0);
  inputData.set(segBytes, initBytes.byteLength);

  // In-memory virtual seekable input stream
  let inPos = 0;

  // In-memory virtual seekable output buffer
  let outCapacity = 64 * 1024;
  let outBuffer = new Uint8Array(outCapacity);
  let outPos = 0;
  let outSize = 0;

  const session = new Mp4DecryptSession();
  try {
    session.addKey(kidHex, keyHex);
    session.setLogLevel(Mp4LogLevel.DEBUG);
    session.onLog((level, msg) => {
      console.log(`  [Native Log ${level}] ${msg}`);
    });

    console.log("\nStarting streaming decryption via decryptCustom()...");
    session.decryptCustom({
      input: {
        read: (buf, count) => {
          const available = Math.min(count, inputData.length - inPos);
          if (available <= 0) return 0;
          buf.set(inputData.subarray(inPos, inPos + available));
          inPos += available;
          return available;
        },
        seek: (offset) => {
          if (offset > inputData.length) return -1;
          inPos = offset;
          return 0;
        },
        tell: () => inPos,
        size: () => inputData.length,
      },
      output: {
        write: (buf, count) => {
          if (outPos + count > outCapacity) {
            outCapacity = Math.max(outCapacity * 2, outPos + count + 4096);
            const newBuf = new Uint8Array(outCapacity);
            newBuf.set(outBuffer);
            outBuffer = newBuf;
          }
          outBuffer.set(buf.subarray(0, count), outPos);
          outPos += count;
          if (outPos > outSize) outSize = outPos;
          return count;
        },
        seek: (offset) => {
          outPos = offset;
          if (outPos > outSize) outSize = outPos;
          return 0;
        },
        tell: () => outPos,
        size: () => outSize,
      },
    });

    const decryptedOutput = outBuffer.subarray(0, outSize);
    console.log(`\nStreamed decrypted output size: ${(decryptedOutput.length / 1024).toFixed(1)} KB`);

    const probeInfo = Mp4DecryptSession.probe(decryptedOutput);
    console.log(`Verification: Track encrypted = ${probeInfo.tracks[0]?.isEncrypted}`);

    if (probeInfo.tracks[0]?.isEncrypted === false) {
      console.log("\nSUCCESS: Streaming callback I/O decrypted media successfully!");
    }
  } finally {
    session.destroy();
  }
}

main().catch(console.error);

File: examples/python/demo_streaming.py
#!/usr/bin/env python3
"""
mp4decrypt-api - Custom Streaming / Callback I/O Demo (Python)
==============================================================

Demonstrates Paradigm 3: Decrypting directly between Python BinaryIO streams
(such as io.BytesIO, network sockets, or custom file wrappers) via C callbacks.

Run with:
    python examples/python/demo_streaming.py
"""

import io
import os
import sys
import urllib.request

# Ensure sdk/python is importable
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..")))
from sdk.python.mp4decrypt_sdk import Mp4DecryptSession, Mp4LogLevel


class TracingStream(io.BytesIO):
    """A stream wrapper that tracks read/write activity for demonstration."""

    def __init__(self, initial_bytes=b"", name="Stream"):
        super().__init__(initial_bytes)
        self.name = name
        self.total_reads = 0
        self.total_writes = 0

    def read(self, size=-1):
        data = super().read(size)
        self.total_reads += len(data)
        return data

    def write(self, b):
        n = super().write(b)
        self.total_writes += n
        return n


def main():
    print("=" * 67)
    print("   MP4DECRYPT-API - CUSTOM STREAMING CALLBACK I/O (PYTHON)         ")
    print("=" * 67)

    base_url = "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey"
    kid_hex = "9eb4050de44b4802932e27d75083e266"
    key_hex = "166634c675823c235a4a9446fad52e4d"

    print("1. Fetching media chunks...")
    init_bytes = urllib.request.urlopen(f"{base_url}/15/init.mp4").read()
    seg_bytes = urllib.request.urlopen(f"{base_url}/15/0001.m4s").read()
    input_data = init_bytes + seg_bytes

    input_stream = TracingStream(input_data, name="InputPipe")
    output_stream = TracingStream(name="OutputPipe")

    print("\n2. Decrypting through custom Python stream interfaces...")
    with Mp4DecryptSession() as session:
        session.add_key(kid_hex, key_hex)
        session.set_log_level(Mp4LogLevel.DEBUG)
        session.on_log(lambda lvl, msg: print(f"   [Stream Log] {msg}"))

        session.decrypt_custom(input_stream, output_stream)
        stats = session.get_stats()

    decrypted_bytes = output_stream.getvalue()

    print(f"\n3. Stream Statistics:")
    print(f"   Input stream bytes read:   {input_stream.total_reads:,}")
    print(f"   Output stream bytes written: {output_stream.total_writes:,}")
    print(f"   Decrypted stream size:      {len(decrypted_bytes):,} bytes")
    print(f"   Execution duration:        {stats.execution_duration_ms:.2f} ms")
    print(f"   Throughput:                {stats.throughput_mb_per_sec:.2f} MB/s")

    # Verification
    info = Mp4DecryptSession.probe(decrypted_bytes)
    print(f"\n4. Verification:")
    print(f"   Track #1 encrypted: {info.tracks[0].is_encrypted}")

    if not info.tracks[0].is_encrypted:
        print("\nSUCCESS: Custom callback streaming decrypted media successfully!")
    else:
        raise RuntimeError("Decryption verification failed.")


if __name__ == "__main__":
    main()
